node/netprio_test.go

```diff
@@ -0,0 +1,35 @@
+// Copyright (C) 2019-2023 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package node
+
+import (
+	"testing"
+
+	"github.com/algorand/go-algorand/test/partitiontest"
+	"github.com/stretchr/testify/require"
+)
+
+// TestBase64AllocboundSize tests that the base64 encoded size of the Nonce is the same as the allocbound
+// used on the Nonce field in the struct for netprio response messages.
+func TestBase64AllocboundSize(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	node := AlgorandFullNode{}
+	require.Len(t, node.NewPrioChallenge(), netPrioChallengeSizeBase64Encoded)
```

Comment:
This is a good start, but we need to make sure the constants used will always produce the right encoded object, even when the underlying libraries change, since the constants used here are independent of the underlying libraries.
For this, we need to lock down the correctness with these values in a test by setting the min/max values to the fields, encode/decode, and make sure the values remain the same.

Is there an example where before this change, data loss would have happened? Then that will be an ideal test case.

Reply:
`node.NewPrioChallenge` is the actual code that generates the encoded strict that populates the field. 

Space of possible Base64 encodings is not unlimited. The only valid variability is whether they pad with `=` to make the length a multiple of 4. This test is sensitive to that variability as well since possible encodings of the current constant would be either 43 or 44 without and with padding, and I'm doing exact length check instead of checking for `<=`. 

Changing the encoding type or the length of the input byte-slice would be a protocol change anyway and the test would catch it as intended.

Reply:
Just added explicit encode/decode and equality check. Not sure what you mean by setting min/max values.

Reply:
We do not need excessive testing there. We do know [base64 encodes 6 bits to a cha](https://datatracker.ietf.org/doc/html/rfc4648#section-4)r (8 bits) so that overhead is 8/6 = 4/3. `roundUp(32 * 4/3) = 43`. Good enough for me.